src/estimation/include/iDynTree/Estimation/ExtendedKalmanFilter.h

```diff
@@ -0,0 +1,104 @@
+     * The system propagation function f() and measurement model function h() are virtual functions to be defined by
+     * the derived class implementing the EKF.
+     * Similarly, the Jacobians for the linearized system propagation and measurement model are also virtual functions
+     * to be implemented by the derived class.
+     *
+     * The derived class must must set the size for states, inputs and outputs using ekfSetStateSize(), ekfSetInputSize() and ekfSetOutputSize() methods.
+     * Then the derived class must call ekfInit() to resize the vectors and matrices for the EKF.
+     *
+     * Similarly before running the estimator through a loop, it is recommended to set the initial states and variances for measurements, system dynamics and intial states.
+     * This is necessary to avoid any NaN values to be propagated.
+     *
+     * Once, initialized properly, the filter can be run by
+     *  - calling ekfSetInputVector() to set the control inputs and then calling ekfPredict() at each prediction step, and
+     *  - calling ekfSetMeasurementVector() to set the measurements and then calling ekfUpdate() at each update step
+     *
+     * The internal state of the estimator can be obtained by calling ekfGetStates()
+     */
+    class DiscreteExtendedKalmanFilter
+    {
+    protected:
+        DiscreteExtendedKalmanFilter();
+        virtual bool f(const iDynTree::VectorDynSize& x_k,
+                       const iDynTree::VectorDynSize& u_k,
+                       const iDynTree::VectorDynSize& w_k,
+                       iDynTree::VectorDynSize& xhat_k_plus_one) = 0;
+
+        virtual bool h(const iDynTree::VectorDynSize& xhat_k_plus_one,
+                       const iDynTree::VectorDynSize& v_k_plus_one,
+                       iDynTree::VectorDynSize& zhat_k_plus_one) = 0;
+
+        virtual bool computejacobianF(iDynTree::VectorDynSize& x, iDynTree::MatrixDynSize& F) = 0;
+        virtual bool computejacobianH(iDynTree::VectorDynSize& x, iDynTree::MatrixDynSize& H) = 0;
+
+        bool ekfPredict();
+        bool ekfUpdate();
+        bool ekfInit();
+        void ekfReset();
+
+        bool ekfSetMeasurementVector(const iDynTree::Span<double>& y);
+        bool ekfSetInputVector(const iDynTree::Span<double>& u);
+        bool ekfSetInitialState(const iDynTree::Span<double>& x0);
+        bool ekfSetStateCovariance(const iDynTree::Span<double>& P);
+        bool ekfSetSystemNoiseMeanAndCovariance(const iDynTree::Span<double>& w_mean, const iDynTree::Span<double>& Q);
+        bool ekfSetMeasurementNoiseMeanAndCovariance(const iDynTree::Span<double>& v_mean, const iDynTree::Span<double>& R);
+        void ekfSetStateSize(size_t dim_X) { m_dim_X = dim_X; }
+        void ekfSetInputSize(size_t dim_U) { m_dim_U = dim_U; }
+        void ekfSetOutputSize(size_t dim_Y) { m_dim_Y = dim_Y; }
+
+        bool ekfGetStates(iDynTree::Span<double> &x) const;
+//         bool ekfGetStateCovariance(iDynTree::Span<double> &P) const;
```

Comment:
leftover?

Reply:
It could have been an useful method which I was reluctant to code at that time.